File: packages/protocol/events/v1/js/src/manifests.ts
import { isSafeRenderId, PROTOCOL } from "./calls";
import {
	firstUnknown,
	hasOwn,
	isPlainObject,
	ProtocolValueError,
	pointer,
	type ValidationIssue,
	validateStrictJson,
} from "./issue";
import type {
	EventComponentClass,
	EventComponentInstance,
	EventHandlerOptions,
	EventsManifest,
} from "./types";

const HTTP_METHOD = /^[!#$%&'*+.^_`|~0-9A-Z-]+$/;
const REVISION = /^[0-9a-f]{64}$/;
const DESCRIPTOR_FIELDS = new Set([
	"componentClassId",
	"eventHandlers",
	"writableStateFields",
]);
const HANDLER_FIELDS = new Set([
	"httpMethod",
	"usesState",
	"debounceMilliseconds",
	"throttleMilliseconds",
	"latestCallWins",
	"allowBatching",
]);
const INSTANCE_FIELDS = new Set([
	"renderId",
	"componentClassId",
	"stateToken",
	"publicState",
]);
const MANIFEST_FIELDS = new Set([
	"protocol",
	"clientGraphRevision",
	"componentClasses",
	"componentInstances",
]);

const prefixed = (base: string, issue: ValidationIssue): ValidationIssue => ({
	path: base + issue.path,
	category: issue.category,
	message: issue.message,
});

const nonNegativeIntegerIssue = (
	value: unknown,
	path: string,
	name: string,
): ValidationIssue | null => {
	if (typeof value !== "number" || !Number.isInteger(value)) {
		return {
			path,
			category:
				typeof value === "number" && !Number.isFinite(value)
					? "strict_json"
					: "type",
			message: `The ${name} hint must be an integer.`,
		};
	}
	if (value < 0) {
		return {
			path,
			category: "range",
			message: `The ${name} hint must be at least 0.`,
		};
	}
	return null;
};

export const validateHandlerDescriptor = (
	value: unknown,
	path = "",
): ValidationIssue | null => {
	const jsonIssue = validateStrictJson(value, path);
	if (jsonIssue) return jsonIssue;
	if (!isPlainObject(value)) {
		return {
			path,
			category: "type",
			message: "Event-handler hints must be an object.",
		};
	}
	if (!hasOwn(value, "httpMethod")) {
		return {
			path: pointer(path, "httpMethod"),
			category: "required",
			message: "Handler hints require 'httpMethod'.",
		};
	}
	const unknown = firstUnknown(value, HANDLER_FIELDS);
	if (unknown !== null) {
		return {
			path: pointer(path, unknown),
			category: "unknown_field",
			message: "Handler hints have an unknown field.",
		};
	}
	if (typeof value.httpMethod !== "string") {
		return {
			path: pointer(path, "httpMethod"),
			category: "type",
			message: "The HTTP method must be a string.",
		};
	}
	if (!HTTP_METHOD.test(value.httpMethod)) {
		return {
			path: pointer(path, "httpMethod"),
			category: "pattern",
			message: "The HTTP method must be an uppercase token.",
		};
	}
	if (hasOwn(value, "usesState") && value.usesState !== true) {
		return {
			path: pointer(path, "usesState"),
			category: "enum",
			message: "The usesState hint has its non-default literal value.",
		};
	}
	for (const name of [
		"debounceMilliseconds",
		"throttleMilliseconds",
	] as const) {
		if (!hasOwn(value, name)) continue;
		const issue = nonNegativeIntegerIssue(
			value[name],
			pointer(path, name),
			name,
		);
		if (issue) return issue;
	}
	if (hasOwn(value, "latestCallWins") && value.latestCallWins !== true) {
		return {
			path: pointer(path, "latestCallWins"),
			category: "enum",
			message: "The latestCallWins hint has its non-default literal value.",
		};
	}
	if (hasOwn(value, "allowBatching") && value.allowBatching !== false) {
		return {
			path: pointer(path, "allowBatching"),
			category: "enum",
			message: "The allowBatching hint has its non-default literal value.",
		};
	}
	return null;
};

export const validateDescriptor = (
	value: unknown,
	path = "",
): ValidationIssue | null => {
	const jsonIssue = validateStrictJson(value, path);
	if (jsonIssue) return jsonIssue;
	if (!isPlainObject(value)) {
		return {
			path,
			category: "type",
			message: "A component descriptor must be an object.",
		};
	}
	for (const required of ["componentClassId", "eventHandlers"]) {
		if (!hasOwn(value, required)) {
			return {
				path: pointer(path, required),
				category: "required",
				message: `The descriptor requires '${required}'.`,
			};
		}
	}
	const unknown = firstUnknown(value, DESCRIPTOR_FIELDS);
	if (unknown !== null) {
		return {
			path: pointer(path, unknown),
			category: "unknown_field",
			message: "The descriptor has an unknown field.",
		};
	}
	if (typeof value.componentClassId !== "string") {
		return {
			path: pointer(path, "componentClassId"),
			category: "type",
			message: "The component class ID must be a string.",
		};
	}
	if (!value.componentClassId) {
		return {
			path: pointer(path, "componentClassId"),
			category: "range",
			message: "The component class ID must not be empty.",
		};
	}
	if (!isPlainObject(value.eventHandlers)) {
		return {
			path: pointer(path, "eventHandlers"),
			category: "type",
			message: "Event handlers must be an object.",
		};
	}
	for (const name of Object.keys(value.eventHandlers).sort()) {
		if (!name) {
			return {
				path: pointer(path, "eventHandlers"),
				category: "range",
				message: "A handler name must not be empty.",
			};
		}
		const issue = validateHandlerDescriptor(
			value.eventHandlers[name],
			pointer(pointer(path, "eventHandlers"), name),
		);
		if (issue) return issue;
	}
	if (hasOwn(value, "writableStateFields")) {
		const fields = value.writableStateFields;
		const fieldPath = pointer(path, "writableStateFields");
		if (!Array.isArray(fields)) {
			return {
				path: fieldPath,
				category: "type",
				message: "Writable State fields must be an array.",
			};
		}
		const seen = new Set<string>();
		for (let index = 0; index < fields.length; index += 1) {
			const field = fields[index];
			if (typeof field !== "string") {
				return {
					path: pointer(fieldPath, index),
					category: "type",
					message: "A writable State field must be a string.",
				};
			}
			if (!field) {
				return {
					path: pointer(fieldPath, index),
					category: "range",
					message: "A writable State field must not be empty.",
				};
			}
			if (seen.has(field)) {
				return {
					path: fieldPath,
					category: "semantic",
					message: "Writable State fields must be unique.",
				};
			}
			seen.add(field);
		}
	}
	return null;
};

export const validateComponentInstance = (
	value: unknown,
	path = "",
): ValidationIssue | null => {
	const jsonIssue = validateStrictJson(value, path);
	if (jsonIssue) return jsonIssue;
	if (!isPlainObject(value)) {
		return {
			path,
			category: "type",
			message: "A component instance must be an object.",
		};
	}
	for (const required of INSTANCE_FIELDS) {
		if (!hasOwn(value, required)) {
			return {
				path: pointer(path, required),
				category: "required",
				message: `The instance requires '${required}'.`,
			};
		}
	}
	const unknown = firstUnknown(value, INSTANCE_FIELDS);
	if (unknown !== null) {
		return {
			path: pointer(path, unknown),
			category: "unknown_field",
			message: "The component instance has an unknown field.",
		};
	}
	if (typeof value.renderId !== "string") {
		return {
			path: pointer(path, "renderId"),
			category: "type",
			message: "The render ID must be a string.",
		};
	}
	if (!isSafeRenderId(value.renderId)) {
		return {
			path: pointer(path, "renderId"),
			category: "pattern",
			message: "The render ID has invalid characters.",
		};
	}
	if (typeof value.componentClassId !== "string") {
		return {
			path: pointer(path, "componentClassId"),
			category: "type",
			message: "The component class ID must be a string.",
		};
	}
	if (!value.componentClassId) {
		return {
			path: pointer(path, "componentClassId"),
			category: "range",
			message: "The component class ID must not be empty.",
		};
	}
	if (value.stateToken !== null && typeof value.stateToken !== "string") {
		return {
			path: pointer(path, "stateToken"),
			category: "type",
			message: "The state token must be a string or null.",
		};
	}
	if (value.stateToken === "") {
		return {
			path: pointer(path, "stateToken"),
			category: "range",
			message: "The state token must not be empty.",
		};
	}
	if (!isPlainObject(value.publicState)) {
		return {
			path: pointer(path, "publicState"),
			category: "type",
			message: "Public State must be an object.",
		};
	}
	const issue = validateStrictJson(value.publicState);
	return issue ? prefixed(pointer(path, "publicState"), issue) : null;
};

export const validateManifest = (
	value: unknown,
	path = "",
): ValidationIssue | null => {
	const jsonIssue = validateStrictJson(value, path);
	if (jsonIssue) return jsonIssue;
	if (!isPlainObject(value)) {
		return {
			path,
			category: "type",
			message: "The Events manifest must be an object.",
		};
	}
	for (const required of MANIFEST_FIELDS) {
		if (!hasOwn(value, required)) {
			return {
				path: pointer(path, required),
				category: "required",
				message: `The manifest requires '${required}'.`,
			};
		}
	}
	const unknown = firstUnknown(value, MANIFEST_FIELDS);
	if (unknown !== null) {
		return {
			path: pointer(path, unknown),
			category: "unknown_field",
			message: "The manifest has an unknown field.",
		};
	}
	if (value.protocol !== PROTOCOL) {
		return {
			path: pointer(path, "protocol"),
			category: typeof value.protocol === "string" ? "enum" : "type",
			message: "The manifest protocol must be citry-events/1.",
		};
	}
	if (
		value.clientGraphRevision !== null &&
		typeof value.clientGraphRevision !== "string"
	) {
		return {
			path: pointer(path, "clientGraphRevision"),
			category: "type",
			message: "The graph revision must be a string or null.",
		};
	}
	if (
		typeof value.clientGraphRevision === "string" &&
		!REVISION.test(value.clientGraphRevision)
	) {
		return {
			path: pointer(path, "clientGraphRevision"),
			category: "pattern",
			message: "The graph revision must be lowercase SHA-256.",
		};
	}
	if (!Array.isArray(value.componentClasses)) {
		return {
			path: pointer(path, "componentClasses"),
			category: "type",
			message: "Component classes must be an array.",
		};
	}
	for (let index = 0; index < value.componentClasses.length; index += 1) {
		const issue = validateDescriptor(
			value.componentClasses[index],
			pointer(pointer(path, "componentClasses"), index),
		);
		if (issue) return issue;
	}
	if (!Array.isArray(value.componentInstances)) {
		return {
			path: pointer(path, "componentInstances"),
			category: "type",
			message: "Component instances must be an array.",
		};
	}
	for (let index = 0; index < value.componentInstances.length; index += 1) {
		const issue = validateComponentInstance(
			value.componentInstances[index],
			pointer(pointer(path, "componentInstances"), index),
		);
		if (issue) return issue;
	}

	const classes = value.componentClasses as EventComponentClass[];
	const instances = value.componentInstances as EventComponentInstance[];
	const classIds = new Set<string>();
	for (let index = 0; index < classes.length; index += 1) {
		const descriptor = classes[index];
		if (classIds.has(descriptor.componentClassId)) {
			return {
				path: pointer(pointer(path, "componentClasses"), index),
				category: "semantic",
				message: `Duplicate component class ID '${descriptor.componentClassId}'.`,
			};
		}
		classIds.add(descriptor.componentClassId);
	}
	const renderIds = new Set<string>();
	for (let index = 0; index < instances.length; index += 1) {
		const instance = instances[index];
		if (renderIds.has(instance.renderId)) {
			return {
				path: pointer(pointer(path, "componentInstances"), index),
				category: "semantic",
				message: `Duplicate render ID '${instance.renderId}'.`,
			};
		}
		renderIds.add(instance.renderId);
	}
	for (let index = 0; index < instances.length; index += 1) {
		const instance = instances[index];
		if (!classIds.has(instance.componentClassId)) {
			return {
				path: pointer(
					pointer(pointer(path, "componentInstances"), index),
					"componentClassId",
				),
				category: "semantic",
				message: "The instance refers to an unknown component class.",
			};
		}
	}
	for (let index = 0; index < instances.length; index += 1) {
		const instance = instances[index];
		if (
			instance.stateToken === null &&
			Object.keys(instance.publicState).length > 0
		) {
			return {
				path: pointer(
					pointer(pointer(path, "componentInstances"), index),
					"publicState",
				),
				category: "semantic",
				message: "A stateless instance must have empty public State.",
			};
		}
	}
	return null;
};

export const assertValidManifest = (value: unknown): EventsManifest => {
	const issue = validateManifest(value);
	if (issue) throw new ProtocolValueError(issue);
	return value as EventsManifest;
};

export const assertValidDescriptor = (value: unknown): EventComponentClass => {
	const issue = validateDescriptor(value);
	if (issue) throw new ProtocolValueError(issue);
	return value as EventComponentClass;
};

export const assertValidHandlerDescriptor = (
	value: unknown,
): EventHandlerOptions => {
	const issue = validateHandlerDescriptor(value);
	if (issue) throw new ProtocolValueError(issue);
	return value as EventHandlerOptions;
};

File: packages/protocol/events/v1/python/citry_events/manifests.py
"""Descriptor and browser-manifest helpers for citry-events/1."""

from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any

from .calls import PROTOCOL, valid_render_id
from .issues import (
    ProtocolValueError,
    ValidationIssue,
    copy_json,
    first_unknown,
    is_finite_json_number,
    pointer,
    utf16_key,
    validate_strict_json,
)

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence

_HTTP_METHOD = re.compile(r"^[!#$%&'*+.^_`|~0-9A-Z-]+$")
_REVISION = re.compile(r"^[0-9a-f]{64}$")
_DESCRIPTOR_FIELDS = ("componentClassId", "eventHandlers", "writableStateFields")
_HANDLER_FIELDS = (
    "httpMethod",
    "usesState",
    "debounceMilliseconds",
    "throttleMilliseconds",
    "latestCallWins",
    "allowBatching",
)
_INSTANCE_FIELDS = ("renderId", "componentClassId", "stateToken", "publicState")
_MANIFEST_FIELDS = ("protocol", "clientGraphRevision", "componentClasses", "componentInstances")


def validate_handler_descriptor(value: Any, path: str = "") -> ValidationIssue | None:
    """Return the first issue in one handler-hints record."""
    json_issue = validate_strict_json(value, path)
    if json_issue is not None:
        return json_issue
    return _validate_handler_descriptor_shape(value, path)


def _validate_handler_descriptor_shape(value: Any, path: str) -> ValidationIssue | None:
    """Validate handler hints after their containing value passed strict JSON."""
    if not isinstance(value, dict):
        return ValidationIssue(path, "type", "Event-handler hints must be an object.")
    if "httpMethod" not in value:
        return ValidationIssue(pointer(path, "httpMethod"), "required", "Handler hints require 'httpMethod'.")
    found, unknown = first_unknown(value, set(_HANDLER_FIELDS))
    if found:
        return ValidationIssue(pointer(path, unknown), "unknown_field", "Handler hints have an unknown field.")
    method = value["httpMethod"]
    if not isinstance(method, str):
        return ValidationIssue(pointer(path, "httpMethod"), "type", "The HTTP method must be a string.")
    if _HTTP_METHOD.fullmatch(method) is None:
        return ValidationIssue(pointer(path, "httpMethod"), "pattern", "The HTTP method must be an uppercase token.")
    if "usesState" in value and value["usesState"] is not True:
        return ValidationIssue(
            pointer(path, "usesState"), "enum", "The usesState hint has its non-default literal value."
        )
    for name in ("debounceMilliseconds", "throttleMilliseconds"):
        if name not in value:
            continue
        duration = value[name]
        if isinstance(duration, bool) or not isinstance(duration, (int, float)):
            return ValidationIssue(pointer(path, name), "type", f"The {name} hint must be an integer.")
        if not is_finite_json_number(duration):
            return ValidationIssue(
                pointer(path, name), "strict_json", f"The {name} hint is outside the browser JSON range."
            )
        if isinstance(duration, float) and not duration.is_integer():
            return ValidationIssue(pointer(path, name), "type", f"The {name} hint must be an integer.")
        if duration < 0:
            return ValidationIssue(pointer(path, name), "range", f"The {name} hint must be at least 0.")
    if "latestCallWins" in value and value["latestCallWins"] is not True:
        return ValidationIssue(
            pointer(path, "latestCallWins"), "enum", "The latestCallWins hint has its non-default literal value."
        )
    if "allowBatching" in value and value["allowBatching"] is not False:
        return ValidationIssue(
            pointer(path, "allowBatching"), "enum", "The allowBatching hint has its non-default literal value."
        )
    return None


def validate_descriptor(value: Any, path: str = "") -> ValidationIssue | None:
    """Return the first issue in one component-class descriptor."""
    json_issue = validate_strict_json(value, path)
    if json_issue is not None:
        return json_issue
    return _validate_descriptor_shape(value, path)


def _validate_descriptor_shape(value: Any, path: str) -> ValidationIssue | None:
    """Validate one descriptor after its containing value passed strict JSON."""
    if not isinstance(value, dict):
        return ValidationIssue(path, "type", "A component descriptor must be an object.")
    for required in ("componentClassId", "eventHandlers"):
        if required not in value:
            return ValidationIssue(pointer(path, required), "required", f"The descriptor requires {required!r}.")
    found, unknown = first_unknown(value, set(_DESCRIPTOR_FIELDS))
    if found:
        return ValidationIssue(pointer(path, unknown), "unknown_field", "The descriptor has an unknown field.")
    class_id = value["componentClassId"]
    if not isinstance(class_id, str):
        return ValidationIssue(pointer(path, "componentClassId"), "type", "The component class ID must be a string.")
    if not class_id:
        return ValidationIssue(pointer(path, "componentClassId"), "range", "The component class ID must not be empty.")
    handlers = value["eventHandlers"]
    if not isinstance(handlers, dict):
        return ValidationIssue(pointer(path, "eventHandlers"), "type", "Event handlers must be an object.")
    if any(not isinstance(name, str) for name in handlers):
        return ValidationIssue(pointer(path, "eventHandlers"), "strict_json", "A handler name must be a string.")
    for name in sorted(handlers, key=utf16_key):
        if not name:
            return ValidationIssue(pointer(path, "eventHandlers"), "range", "A handler name must not be empty.")
        issue = _validate_handler_descriptor_shape(handlers[name], pointer(pointer(path, "eventHandlers"), name))
        if issue is not None:
            return issue
    if "writableStateFields" in value:
        fields = value["writableStateFields"]
        field_path = pointer(path, "writableStateFields")
        if not isinstance(fields, list):
            return ValidationIssue(field_path, "type", "Writable State fields must be an array.")
        seen: set[str] = set()
        for index, field in enumerate(fields):
            if not isinstance(field, str):
                return ValidationIssue(pointer(field_path, index), "type", "A writable State field must be a string.")
            if not field:
                return ValidationIssue(
                    pointer(field_path, index), "range", "A writable State field must not be empty."
                )
            if field in seen:
                return ValidationIssue(field_path, "semantic", "Writable State fields must be unique.")
            seen.add(field)
    return None


def build_handler_descriptor(
    http_method: str,
    *,
    uses_state: bool = False,
    debounce_milliseconds: int | None = None,
    throttle_milliseconds: int | None = None,
    latest_call_wins: bool = False,
    allow_batching: bool = True,
) -> dict[str, Any]:
    """Build one handler's compact browser hints."""
    hints: dict[str, Any] = {"httpMethod": http_method}
    if uses_state:
        hints["usesState"] = True
    if debounce_milliseconds is not None:
        hints["debounceMilliseconds"] = debounce_milliseconds
    if throttle_milliseconds is not None:
        hints["throttleMilliseconds"] = throttle_milliseconds
    if latest_call_wins:
        hints["latestCallWins"] = True
    if not allow_batching:
        hints["allowBatching"] = False
    issue = validate_handler_descriptor(hints)
    if issue is not None:
        raise ProtocolValueError(issue)
    return hints


def build_descriptor(
    component_class_id: str,
    event_handlers: Mapping[str, Mapping[str, Any]],
    *,
    writable_state_fields: Sequence[str] | None = None,
) -> dict[str, Any]:
    """Build one fresh component-class descriptor."""
    descriptor: dict[str, Any] = {
        "componentClassId": component_class_id,
        "eventHandlers": {name: copy_json(dict(hints)) for name, hints in event_handlers.items()},
    }
    if writable_state_fields is not None:
        descriptor["writableStateFields"] = list(writable_state_fields)
    issue = validate_descriptor(descriptor)
    if issue is not None:
        raise ProtocolValueError(issue)
    return descriptor


def validate_component_instance(value: Any, path: str = "") -> ValidationIssue | None:
    """Return the first issue in one rendered component-instance record."""
    json_issue = validate_strict_json(value, path)
    if json_issue is not None:
        return json_issue
    return _validate_component_instance_shape(value, path)


def _validate_component_instance_shape(value: Any, path: str) -> ValidationIssue | None:
    """Validate one instance after its containing value passed strict JSON."""
    if not isinstance(value, dict):
        return ValidationIssue(path, "type", "A component instance must be an object.")
    for required in _INSTANCE_FIELDS:
        if required not in value:
            return ValidationIssue(pointer(path, required), "required", f"The instance requires {required!r}.")
    found, unknown = first_unknown(value, set(_INSTANCE_FIELDS))
    if found:
        return ValidationIssue(pointer(path, unknown), "unknown_field", "The component instance has an unknown field.")
    render_id = value["renderId"]
    if not isinstance(render_id, str):
        return ValidationIssue(pointer(path, "renderId"), "type", "The render ID must be a string.")
    if not valid_render_id(render_id):
        return ValidationIssue(pointer(path, "renderId"), "pattern", "The render ID has invalid characters.")
    class_id = value["componentClassId"]
    if not isinstance(class_id, str):
        return ValidationIssue(pointer(path, "componentClassId"), "type", "The component class ID must be a string.")
    if not class_id:
        return ValidationIssue(pointer(path, "componentClassId"), "range", "The component class ID must not be empty.")
    token = value["stateToken"]
    if token is not None and not isinstance(token, str):
        return ValidationIssue(pointer(path, "stateToken"), "type", "The state token must be a string or null.")
    if token == "":
        return ValidationIssue(pointer(path, "stateToken"), "range", "The state token must not be empty.")
    public_state = value["publicState"]
    if not isinstance(public_state, dict):
        return ValidationIssue(pointer(path, "publicState"), "type", "Public State must be an object.")
    return None


def build_component_instance(
    render_id: str,
    component_class_id: str,
    state_token: str | None,
    public_state: Mapping[str, Any],
) -> dict[str, Any]:
    """Build one fresh component-instance record."""
    instance = {
        "renderId": render_id,
        "componentClassId": component_class_id,
        "stateToken": state_token,
        "publicState": copy_json(dict(public_state)),
    }
    issue = validate_component_instance(instance)
    if issue is not None:
        raise ProtocolValueError(issue)
    return instance


def validate_manifest(value: Any, path: str = "") -> ValidationIssue | None:
    """Return the first structural or relationship issue in a browser manifest."""
    json_issue = validate_strict_json(value, path)
    if json_issue is not None:
        return json_issue
    return _validate_manifest_shape(value, path)


def _validate_manifest_shape(value: Any, path: str) -> ValidationIssue | None:
    """Validate one manifest after the complete value passed strict JSON."""
    if not isinstance(value, dict):
        return ValidationIssue(path, "type", "The Events manifest must be an object.")
    for required in _MANIFEST_FIELDS:
        if required not in value:
            return ValidationIssue(pointer(path, required), "required", f"The manifest requires {required!r}.")
    found, unknown = first_unknown(value, set(_MANIFEST_FIELDS))
    if found:
        return ValidationIssue(pointer(path, unknown), "unknown_field", "The manifest has an unknown field.")
    if value["protocol"] != PROTOCOL:
        category = "type" if not isinstance(value["protocol"], str) else "enum"
        return ValidationIssue(pointer(path, "protocol"), category, "The manifest protocol must be citry-events/1.")
    revision = value["clientGraphRevision"]
    if revision is not None and not isinstance(revision, str):
        return ValidationIssue(
            pointer(path, "clientGraphRevision"), "type", "The graph revision must be a string or null."
        )
    if isinstance(revision, str) and _REVISION.fullmatch(revision) is None:
        return ValidationIssue(
            pointer(path, "clientGraphRevision"), "pattern", "The graph revision must be lowercase SHA-256."
        )
    classes = value["componentClasses"]
    if not isinstance(classes, list):
        return ValidationIssue(pointer(path, "componentClasses"), "type", "Component classes must be an array.")

    for index, descriptor in enumerate(classes):
        item_path = pointer(pointer(path, "componentClasses"), index)
        issue = _validate_descriptor_shape(descriptor, item_path)
        if issue is not None:
            return issue

    instances = value["componentInstances"]
    if not isinstance(instances, list):
        return ValidationIssue(pointer(path, "componentInstances"), "type", "Component instances must be an array.")
    for index, instance in enumerate(instances):
        item_path = pointer(pointer(path, "componentInstances"), index)
        issue = _validate_component_instance_shape(instance, item_path)
        if issue is not None:
            return issue

    class_ids: set[str] = set()
    for index, descriptor in enumerate(classes):
        item_path = pointer(pointer(path, "componentClasses"), index)
        class_id = descriptor["componentClassId"]
        if class_id in class_ids:
            return ValidationIssue(item_path, "semantic", f"Duplicate component class ID {class_id!r}.")
        class_ids.add(class_id)
    render_ids: set[str] = set()
    for index, instance in enumerate(instances):
        item_path = pointer(pointer(path, "componentInstances"), index)
        render_id = instance["renderId"]
        if render_id in render_ids:
            return ValidationIssue(item_path, "semantic", f"Duplicate render ID {render_id!r}.")
        render_ids.add(render_id)
    for index, instance in enumerate(instances):
        item_path = pointer(pointer(path, "componentInstances"), index)
        if instance["componentClassId"] not in class_ids:
            return ValidationIssue(
                pointer(item_path, "componentClassId"),
                "semantic",
                "The instance refers to an unknown component class.",
            )
    for index, instance in enumerate(instances):
        item_path = pointer(pointer(path, "componentInstances"), index)
        if instance["stateToken"] is None and instance["publicState"]:
            return ValidationIssue(
                pointer(item_path, "publicState"), "semantic", "A stateless instance must have empty public State."
            )
    return None


def build_manifest(
    client_graph_revision: str | None,
    component_classes: Sequence[Mapping[str, Any]],
    component_instances: Sequence[Mapping[str, Any]],
) -> dict[str, Any]:
    """Build a fresh, fully validated Events browser manifest."""
    manifest = {
        "protocol": PROTOCOL,
        "clientGraphRevision": client_graph_revision,
        "componentClasses": [copy_json(dict(item)) for item in component_classes],
        "componentInstances": [copy_json(dict(item)) for item in component_instances],
    }
    issue = validate_manifest(manifest)
    if issue is not None:
        raise ProtocolValueError(issue)
    return manifest
